api/billingd/service/service.go

```diff
@@ -0,0 +1,708 @@
+
+	r := s.db.FindOne(ctx, bson.M{"_id": req.CustomerId})
+	if r.Err() != nil {
+		return nil, r.Err()
+	}
+	var doc Customer
+	if err := r.Decode(&doc); err != nil {
+		return nil, err
+	}
+	log.Debugf("got customer %s", doc.ID)
+	return &pb.GetCustomerResponse{
+		Email:          doc.Email,
+		Status:         doc.Status,
+		Balance:        doc.Balance,
+		Billable:       doc.Billable,
+		Delinquent:     doc.Delinquent,
+		CreatedAt:      doc.CreatedAt,
+		Period:         periodToPb(doc.Period),
+		StoredData:     usageToPb(doc.StoredData, StoredDataUnitSize, StoredDataFreeUnits),
+		NetworkEgress:  usageToPb(doc.NetworkEgress, NetworkEgressUnitSize, NetworkEgressFreeUnits),
+		InstanceReads:  usageToPb(doc.InstanceReads, InstanceReadsUnitSize, InstanceReadsFreeUnits),
+		InstanceWrites: usageToPb(doc.InstanceWrites, InstanceWritesUnitSize, InstanceWritesFreeUnits),
+	}, nil
+}
+
+func periodToPb(period Period) *pb.Period {
+	return &pb.Period{
+		Start: period.Start,
+		End:   period.End,
+	}
+}
+
+func usageToPb(usage Usage, unitSize, freeUnits int64) *pb.Usage {
+	free := (freeUnits * unitSize) - usage.Total
+	if free < 0 {
+		free = 0
+	}
+	return &pb.Usage{
+		Units: usage.Units,
+		Total: usage.Total,
+		Free:  free,
+	}
+}
+
+func (s *Service) GetCustomerSession(_ context.Context, req *pb.GetCustomerSessionRequest) (
+	*pb.GetCustomerSessionResponse, error) {
+	lck := s.semaphores.Get(customerLock(req.CustomerId))
+	lck.Acquire()
+	defer lck.Release()
+
+	session, err := s.stripe.BillingPortalSessions.New(&stripe.BillingPortalSessionParams{
+		Customer:  stripe.String(req.CustomerId),
+		ReturnURL: stripe.String(s.config.StripeSessionReturnURL),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &pb.GetCustomerSessionResponse{
+		Url: session.URL,
+	}, nil
+}
+
+func (s *Service) UpdateCustomer(ctx context.Context, req *pb.UpdateCustomerRequest) (
+	*pb.UpdateCustomerResponse, error) {
+	lck := s.semaphores.Get(customerLock(req.CustomerId))
+	lck.Acquire()
+	defer lck.Release()
+
+	if _, err := s.db.UpdateOne(ctx, bson.M{"_id": req.CustomerId}, bson.M{
+		"$set": bson.M{"balance": req.Balance, "billable": req.Billable, "delinquent": req.Delinquent},
+	}); err != nil {
+		return nil, err
+	}
+	log.Debugf("updated customer %s", req.CustomerId)
+	return &pb.UpdateCustomerResponse{}, nil
+}
+
+func (s *Service) UpdateCustomerSubscription(ctx context.Context, req *pb.UpdateCustomerSubscriptionRequest) (
+	*pb.UpdateCustomerSubscriptionResponse, error) {
+	lck := s.semaphores.Get(customerLock(req.CustomerId))
+	lck.Acquire()
+	defer lck.Release()
+
+	r := s.db.FindOneAndUpdate(ctx, bson.M{"_id": req.CustomerId}, bson.M{
+		"$set": bson.M{
+			"status":       req.Status,
+			"period.start": req.Period.Start,
+			"period.end":   req.Period.End,
+		},
+	})
+	if r.Err() != nil {
+		return nil, r.Err()
+	}
+	var pre Customer
+	if err := r.Decode(&pre); err != nil {
+		return nil, err
+	}
+	if pre.Period.End < req.Period.End {
```

Comment:
If the update includes a period change, we reset the time-based usage stats.

Reply:
The period changes whenever the billing cycle ends, which is currently every day. So, this should happen each day.